ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/deduplication/CommandDeduplicationBase.scala

```diff
@@ -315,4 +301,76 @@ private[testtool] abstract class CommandDeduplicationBase(
       )
     }
   })
+
+  def requestHasOkCompletion(
+      ledger: ParticipantTestContext
+  )(request: SubmitRequest, party: Party)(implicit ec: ExecutionContext): Future[Completion] = {
+    requestHasCompletionWithStatusCode(ledger)(request, party, Code.OK)
+  }
+
+  def requestHasCompletionWithStatusCode(
+      ledger: ParticipantTestContext
+  )(request: SubmitRequest, party: Party, code: Code)(implicit
+      ec: ExecutionContext
+  ): Future[Completion] = {
+    requestIsSubmittedAndHasCompletion(ledger)(request, party).map(completion => {
+      assert(
+        completion.getStatus.code == code.value(),
+        s"Expecting completion with status code $code but completion has status ${completion.status}",
+      )
+      completion
+    })
+  }
+
+  protected def requestFailsWithStatus(
+      ledger: ParticipantTestContext
+  )(request: SubmitRequest, statusCode: Code)(implicit ec: ExecutionContext): Future[Unit] = {
+    ledger
+      .submit(request)
+      .mustFail(s"Request expected to fail with status $statusCode")
+      .map(assertGrpcError(_, statusCode, None))
+  }
+
+  protected def requestIsSubmittedAndHasCompletion(
```

Comment:
Also, I would consider naming the functions like `submitRequestAnd...`, for example `submitRequestAndAssertCompletion`. Or maybe it would be worth splitting up to preserve the SRP(?)

Reply:
I would split it but they're almost always used together so it's quite handy. Agree on the name change, will rename

Reply:
We don't really assert the completion though, so I would rename it to `submitRequestAndFindCompletion`